snuba/datasets/tags_column_processor.py

```diff
@@ -16,6 +27,39 @@
 NESTED_COL_EXPR_RE = re.compile(r"^([a-zA-Z0-9_\.]+)\[([a-zA-Z0-9_\.:-]+)\]$")
 
 
+@dataclass(frozen=True)
+class ParsedNestedColumn:
+    """
+    Provides some structure to the tags/contexts column to avoid parsing the string
+    in multiple places around the code.
+    """
+
+    col_name: str
+    parameter: Optional[str]
+
+    @classmethod
+    def parse_column_expression(cls, col_expr: str) -> Optional[ParsedNestedColumn]:
+        match = NESTED_COL_EXPR_RE.match(col_expr)
+        if match:
+            col_prefix = match[1]
+            param_string = match[2]
+
+            if col_prefix in ("tags", "contexts"):
+                assert param_string
```

Comment:
When would you expect this condition block not to be entered? I also don't see how parameter would ever be falsey here (it can't be None or an empty string, per the regular expression.)

Reply:
Looking at the changes in f59a3ac4d32433c1456a6bfc9f8f2fc31967036f (and probably just having some time to absorb this change in general) this data structure sort of makes more sense to me now. I think part of my confusion comes from how this class is used to represent both the subscript form (`tags[xyz]`) and non-subscript form (`tags_key`) like noted in https://github.com/getsentry/snuba/pull/707#discussion_r372126754.

I still don't understand the purpose of [this assertion](https://github.com/getsentry/snuba/blob/f59a3ac4d32433c1456a6bfc9f8f2fc31967036f/snuba/datasets/tags_column_processor.py#L55-L56) though, considering it is already enforced by the form of the regular expression itself (unless this is just to protect against the potential of the regular expression changing, which seems like it would materially change the meaning/purpose of the regular expression anyway.)

If this is just to ensure that the data structure itself is valid (regardless of the way it was constructed) it seems like that should be in some sort of `__post_init__` validation, instead of in this condition in this alternate constructor that seems impossible to actually reach (unless I forgot how to read regular expressions, or am missing the point entirely.)

Reply:
no, you are right, in the old implementation I made the regex more permissive, I forgot that I reverted that change. The assertion now would never be broken.